# <center>高级心理统计 (Advanced Statistics in Psych Sci) </center>  
## <center>《贝叶斯统计及其在Python中的实现》 (Bayesian inference with Python) </center>  
## <center>Instructor： 胡传鹏（博士）(Dr. Hu Chuan-Peng) </center>  
### <center>南京师范大学心理学院 (School of Psychology, Nanjing Normal University)  </center>

## Outlines  
### 1. 为什么要学习本课程 [Why Bayesia inference]  
### 2. 本课程的内容将是什么 [What is the syllabus]  
###  3. 如何学好这门课[How can I learn this course well]

##  0. 贝叶斯统计课程选课概况  
- 学生身份  

![Image Name](https://cdn.kesci.com/upload/sjboycorpy.png?imageView2/0/w/960/h/960)  

- 教育程度  


![Image Name](https://cdn.kesci.com/upload/sjbp057ve3.png?imageView2/0/w/960/h/960)  


- 专业领域  

![Image Name](https://cdn.kesci.com/upload/sjbor7adly.png?imageView2/0/w/960/h/960)  
  

- 编程经验  

![Image Name](https://cdn.kesci.com/upload/sjbp0nk57u.png?imageView2/0/w/960/h/960)  



### <center>总结：贝叶斯统计线上课程吸引了众多高学历在读学生，但学员普遍缺乏编程基础。<center>

<center><span style="font-size: 72px;">Part 1: Why Bayesian statistics</span></center>

<center> <span style="font-size: 56px;">研究人类心理与行为的规律，  
	容易吗？  </center>  
	

##  1.  为什么要学习本课程 (Why Bayesia inference)  

### 1.1  为什么心理学需要更好的方法 (Why does psychological science need better methods?)  

#### 原因1:  人类认知与行为本身的复杂性

 ![Image Name](https://cdn.kesci.com/upload/image/rhdcyu860w.gif?imageView2/0/w/960/h/960)  

Source: https://www.science.org/toc/science/309/5731

### **Q1:  What is the Uiverse Made of (physics)**  

### **Q2:  What is the Biological Basis of Consciouness (psychological/cognitive science)**

### Q:   重要性和复杂度相似的问题，是否意味着研究方法也应该类似地复杂？

### 物理学中的方法 (Methods in Physics):  

#### Example 1: Webb telescope (韦伯望远镜)  (**equipment**)  
![Image Name](https://cdn.kesci.com/upload/image/rhdd0r46k3.png?imageView2/0/w/720/h/640)  

### Example 2: Big-team science (CERN, the European Organization for Nuclear Research) ---- **equipment & practices**  

### Example 3: **Mathematics**

### 其他研究人类智能的领域所采用的方法 (Methods in other fields that also study "intelligence")  

## **AI**  

![Image Name](https://cdn.kesci.com/upload/image/rhdd1sr5y2.png?imageView2/0/w/640/h/640)  


### 心理科学的研究方法 [What do psychological scientists have?]  
##### 你们能够想到的研究方法包括哪些？  

![Image Name](https://cdn.kesci.com/upload/image/rhdd2dgwc8.png?imageView2/0/w/640/h/640)  

#### **实证研究：**  
* 质性研究  
* 观察法  
* 问卷  
* 行为实验  
* 眼动、生理数据记录  
* EEG/ERP/MEG  
* fMRI/PET/fNIRs  
* TMS/tDCS  
* ...

#### **统计方法：**  
* *t*-test  
* ANOVA  
* Correlation  
* Structural equation model (SEM)  
* ?

#### 相关方法课程：  
* 心理测量  
* 心理统计（包括SPSS等）  
* 实验心理学（包括E-prime等）  
* ？

* 更好的仪器  
* **更好的统计/数据分析**  
* 更好的实践 (e.g., big-team science)

#### 原因2: 更复杂的数据  

* 数据字化的时代，大数据  
* 神经成像/生理数据  
* 多模态的数据融合

  ##  <center>1.2 确实有更好的统计方法<center>

### <cenrer>贝叶斯统计 (Bayesian inference) <cenrer>  


![Image Name](https://cdn.kesci.com/upload/image/rhdf3bb12c.png?imageView2/0/w/640/h/640)  



* 灵活/强大/能用  
* 易用  
* 可拓展性强  
* 方便跨学科交流  
* ...

#### 通用/灵活/强大  

* 通用：贝叶斯推断在多个学科特别是AI领域得到了广泛应用。  
* 灵活：结合先验知识与新数据，灵活调整模型以适应不同数据情境。  
* 强大：与频率主义学派相比，贝叶斯方法适用范围更广。

#### （相对）易用  

概率编程语言(Probabilistic Programming Languages)的发展和普及  


PPLs: *computational languages for statistical modeling*  

* Python	 
  * PyMC  
  * NumPyro   
* R  
  * Stan  
  * JAGS  
* BUGS  
* Julia  
  * Mamba  
  * Turing.jl  
* ...

##### 大部分情况下，开发者使用它可以轻松地定义概率模型，然后程序会自动地求解模型。  


![Image Name](https://cdn.kesci.com/upload/image/rhdf4r9fbh.png?imageView2/0/w/640/h/640)  


Source: https://towardsdatascience.com/intro-to-probabilistic-programming-b47c4e926ec5  


#### 可拓展  

贝叶斯概念已经应用到以深度学习为中心的新技术的发展，包括深度学习框架(TensorFlow, Pytorch)，创建表示能力更强、数据驱动的模型

#### 方便交流  
大部分PPLs都有类似的数据结构，但是不同的学科使用的语言不同。  

心理学/社会科学/神经科学：  
* **PyMC**  (bambi)  
* Stan (brms)  
* BUGS

<center><span style="font-size: 72px;">Part 2: Three examples </span></center>

### 例1：社会关系地位与幸福感的关系  

实例的数据来自[ Many Labs 2 项目](osf.io/uazdm/)中的一个研究。  

该研究探究了社会关系地位对于幸福感的影响 “Sociometric status and well-being”， (Anderson, Kraus, Galinsky, & Keltner, 2012)。  

该数据集包括6905个被试的数据。  

<p style="font-size:14px">参考文献：</p>  
<p style="font-size:14px">  
Anderson, C., Kraus, M. W., Galinsky, A. D., & Keltner, D. (2012). The local-ladder effect: Social status and subjective well-being. Psychological Science, 23(7), 764-771. https://doi.org/10.1177/0956797611434537  
</p>

In [1]:
import arviz as az
from scipy.stats import levene
from pingouin import ttest, tost, bayesfactor_ttest
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pymc as pm
import xarray as xr
import pytensor.tensor as pt
import scipy.stats as stats
import seaborn as sns
from pytensor.tensor import TensorVariable
from scipy.stats import pearsonr
import graphviz
from graphviz import Digraph

# 设置随机种子
rng = np.random.default_rng(202409)

# 设置APA 7的画图样式
plt.rcParams.update({
    'figure.figsize': (4, 3),      # 设置画布大小
    'font.size': 12,               # 设置字体大小
    'axes.titlesize': 12,          # 标题字体大小
    'axes.labelsize': 12,          # 轴标签字体大小
    'xtick.labelsize': 12,         # x轴刻度字体大小
    'ytick.labelsize': 12,         # y轴刻度字体大小
    'lines.linewidth': 1,          # 线宽
    'axes.linewidth': 1,           # 轴线宽度
    'axes.edgecolor': 'black',     # 设置轴线颜色为黑色
    'axes.facecolor': 'white',     # 轴背景颜色（白色）
    'xtick.direction': 'in',       # x轴刻度线向内
    'ytick.direction': 'out',    # y轴刻度线向内和向外
    'xtick.major.size': 6,         # x轴主刻度线长度
    'ytick.major.size': 6,         # y轴主刻度线长度
    'xtick.minor.size': 4,         # x轴次刻度线长度（如果启用次刻度线）
    'ytick.minor.size': 4,         # y轴次刻度线长度（如果启用次刻度线）
    'xtick.major.width': 1,        # x轴主刻度线宽度
    'ytick.major.width': 1,        # y轴主刻度线宽度
    'xtick.minor.width': 0.5,      # x轴次刻度线宽度（如果启用次刻度线）
    'ytick.minor.width': 0.5,      # y轴次刻度线宽度（如果启用次刻度线）
    'ytick.labelleft': True,       # y轴标签左侧显示
    'ytick.labelright': False      # 禁用y轴标签右侧显示
})

In [2]:
# 导入数据, 注意在和鲸和在自己电脑（本地）时有差异）
try:
  SMS_data = pd.read_csv('/home/mw/input/bayes3797/SMS_Well_being.csv')
except:
  SMS_data = pd.read_csv('data/SMS_Well_being.csv')
# 选择需要的列
SMS_data = SMS_data[['uID','variable','factor','Country']]
# 查看数据
SMS_data.head(2)

uID  variable factor Country
0    2 -0.366739   High  Brazil
1    4 -0.917172   High  Brazil

In [3]:
# 把数据分为高低两种社会关系的地位的子数据以便画图与后续分析
factor_high = [sorted(SMS_data.query('factor=="High"').variable.loc[0:3000])]
factor_low = [sorted(SMS_data.query('factor=="Low"').variable.loc[0:3000])]

#### 通过画图可视化两种社会关系地位对幸福感的影响  

图中横坐标代表高低两种社会关系地位，纵坐标代表了主观幸福感评分。

In [4]:
# 定义绘图函数
def adjacent_values(vals, q1, q3):
    upper_adjacent_value = q3 + (q3 - q1) * 1.5
    upper_adjacent_value = np.clip(upper_adjacent_value, q3, vals[-1])

    lower_adjacent_value = q1 - (q3 - q1) * 1.5
    lower_adjacent_value = np.clip(lower_adjacent_value, vals[0], q1)
    return lower_adjacent_value, upper_adjacent_value

def set_axis_style(ax, labels):
    ax.xaxis.set_tick_params(direction='out')
    ax.xaxis.set_ticks_position('bottom')
    ax.set_xticks(np.arange(1, len(labels) + 1), labels=labels)
    ax.set_xlim(0.25, len(labels) + 0.75)
    ax.set_xlabel('Soceity Status')

def plot_violin_apa7(data, x_labels, y_label, title=None):
    
    fig, ax = plt.subplots(figsize=(6, 4))  # 调整图形大小

    # 创建小提琴图
    parts = ax.violinplot(data, showmeans=False, showmedians=False, showextrema=False)

    # 设置小提琴图的样式
    for pc in parts['bodies']:
        pc.set_facecolor('#D43F3A')
        pc.set_edgecolor('black')
        pc.set_alpha(0.8)

    # 计算分位数和须
    quartile1, medians, quartile3 = np.percentile(data, [25, 50, 75], axis=1)
    whiskers = np.array([adjacent_values(sorted_array, q1, q3)
                         for sorted_array, q1, q3 in zip(data, quartile1, quartile3)])

    whiskers_min, whiskers_max = whiskers[:, 0], whiskers[:, 1]

    # 绘制中位数和须
    inds = np.arange(1, len(medians) + 1)
    ax.scatter(inds, medians, marker='o', color='white', s=30, zorder=3)
    ax.vlines(inds, quartile1, quartile3, color='k', linestyle='-', lw=5)
    ax.vlines(inds, whiskers_min, whiskers_max, color='k', linestyle='-', lw=1)

    # 设置x轴和y轴标签
    ax.set_xticks(np.arange(1, len(x_labels) + 1))
    ax.set_xticklabels(x_labels)
    ax.set_ylabel(y_label)

    # 设置y轴从0开始
    ax.set_ylim(0, None)

    # 保留左侧和底部的框线，移除右侧和顶部的框线
    ax.spines['bottom'].set_visible(True)
    ax.spines['left'].set_visible(True)
    ax.spines['bottom'].set_linewidth(1)
    ax.spines['left'].set_linewidth(1)

    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)

    # 如果有标题，则设置标题
    if title:
        ax.set_title(title, pad=10)

    # 设置x轴风格
    set_axis_style(ax, x_labels)

    # 图形的显示
    plt.tight_layout()
    plt.show()

In [5]:
# 示例数据和函数调用
plot_data = [np.random.normal(70, 10, 100), np.random.normal(60, 15, 100)]
plot_violin_apa7(plot_data, ['Low', 'High'], 'Well-being',
                 'The impact of social status on well-being')


<Figure size 600x400 with 1 Axes>

#### 通过*t*检验，分析两种社会关系地位下幸福感的差异  

In [6]:
# levene检验
levene_stat, levene_p = levene(SMS_data['variable'][SMS_data['factor'] == 'High'], 
                               SMS_data['variable'][SMS_data['factor'] == 'Low'])
print(f'Levene Test p-value: {levene_p}')

Levene Test p-value: 0.1560218938980748


根据独立样本*t*检验结果，低社会地位组（*M* = 0.014, *SD* = 0.66）和高社会地位组（*M* = -0.014, *SD* = 0.67）之间的差异在统计学上不显著，*t* = 1.76, *p* = 0.079。  

这意味着我们无法在传统显著性水平（α = 0.05）下拒绝原假设($H_0$)。因此高社会地位和低社会地位下的幸福感差异在统计学上不是显著的。

In [7]:
SMS_low = SMS_data.query('factor=="Low"').variable.values
SMS_high = SMS_data.query('factor=="High"').variable.values
print(
    f"Low Social Status：{np.around(np.mean(SMS_low),3)} ± {np.around(np.std(SMS_low),2)}；",
    f"High Social Status：{np.around(np.mean(SMS_high),3)} ± {np.around(np.std(SMS_high),2)}")

# 进行独立样本t检验  
results = stats.ttest_ind(
    a= SMS_low,
    b= SMS_high, 
    equal_var=True)
print(f"t={round(results[0],2)}, p={round(results[1],3)}")

Low Social Status：0.014 ± 0.66； High Social Status：-0.014 ± 0.67
t=1.76, p=0.079


TOST检验是一种常用于等效性检验的方法，其目的是评估两个组之间的差异是否可以被视为“等效”。在该方法中，等效性被定义为组间差异位于预先设定的“等效边界”范围内。如果组间的差异落在这一范围内，则可以认为这两个组在统计上是等效的。  

等效边界设定为0.2（bound=0.2），意味着如果两个组之间的差异不超过0.2，就可以认为它们是等效的。  

因此，检验得出的*p*值（2.265256e-27）可以表明有非常强的证据表明这两个组之间的差异在等效边界之内。

In [8]:
# 双单侧检验（Two one-side test, TOST）
tost_res = tost(x=SMS_data['variable'][SMS_data['factor'] == 'High'], 
                y=SMS_data['variable'][SMS_data['factor'] == 'Low'], 
                bound=0.20, paired=False) # 将等效边界值设为0.2
print(tost_res)

      bound   dof          pval
TOST    0.2  6903  2.265256e-27


#### 通过贝叶斯因子替代*t*检验  
传统的零假设显著性检验（Null hypothesis significance test, NHST）的框架之下，*t*检验只提供了一个二分的结果：拒绝或者无法拒绝$H_0$。 但 ***p* = 0.079**这样的结果无法支持$H_0$。而贝叶斯因子（Bayes Factor, BF）方法可以提供一种更直观和渐进的方式来评估数据支持原假设（无效假设，$H_0$）还是备择假设($H_1$)。 

In [9]:
# 贝叶斯因子
## 比较高低社会关系地位的社会关系的效应
ttest_res = ttest(SMS_data['variable'][SMS_data['factor'] == 'High'], 
                     SMS_data['variable'][SMS_data['factor'] == 'Low'], 
                     paired=False)

## 计算贝叶斯因子
t_value = ttest_res['T'].values[0] # 提取t值
dof = ttest_res['dof'].values[0] # 提取自由度

# 计算‘High’组和‘Low’组的样本量大小
sample_size_high = len(SMS_data['variable'][SMS_data['factor'] == 'High'])
sample_size_low = len(SMS_data['variable'][SMS_data['factor'] == 'Low'])

# 使用默认的r值= 0.707
BF_sms = bayesfactor_ttest(t_value, nx=sample_size_high, ny=sample_size_low, paired=False, r=0.707)

# 输出贝叶斯因子
print(BF_sms)

0.12703440686826656


贝叶斯因子为$BF_{10}$ = 0.127表明，对于当前数据支持零假设$H_0$ ，说明高低社会关系地位组之间的社会关系效应没有显著差异。  

贝叶斯因子（Bayes Factor，$BF$）  

| 贝叶斯因子，$BF_{01}$ | 解释标准                      |  
|-----------------------------|-------------------------------|  
| > 100                        | 极强的证据支持$H_0$    |  
| 30 ~ 100                     | 非常强的证据支持$H_0$ |  
| 10 ~ 30                      | 较强的证据支持$H_0$    |  
| 3 ~ 10                       | 中等程度的证据支持$H_0$ |  
| 1 ~ 3                        | 较弱的证据支持$H_0$    |  
| 1                            | 没有证据                      |  
| 1/3 ~ 1                      | 较弱的证据支持$H_1$    |  
| 1/10 ~ 1/3                   | 中等程度的证据支持$H_1$|  
| 1/30 ~ 1/10                  | 较强的证据支持$H_1$    |  
| 1/100 ~ 1/30                 | 非常强的证据支持$H_1$  |  
| < 1/100                      | 极强的证据支持$H_1$    |  

<p style="font-size:14px">参考文献：</p>  

<p style="font-size:14px">  
胡传鹏, 孔祥祯, Wagenmakers, E.-J., Ly, A., 彭凯平 (2018). 贝叶斯因子及其在JASP中的实现. 心理科学进展, 26(6), 951-965.  
</p>  

<p style="font-size:14px">  
Wagenmakers, E.-J., & Brown, S. (2007). On the linear relation between the mean and the standard deviation of a response time distribution. Psychological Review, 114(3), 830-841. https://doi.org/10.1037/0033-295X.114.3.830  
</p>

In [10]:
# 敏感性分析
# Cauchy (0, 1)
BF_sms_1 = bayesfactor_ttest(t_value, nx=sample_size_high, ny=sample_size_low, paired=False, r=1)
print(BF_sms_1)

0.09002474489275247


### 贝叶斯推断  


这个模型是怎么来的呢？  

一个简单的线性模型：  

1. 通过建立线性模型去替代原本的*t*检验模型。  

2. 通过PyMC对后验进行采样  

3. 通过Arviz对结果进行展示，辅助统计推断

In [11]:
# 通过pymc建立基于贝叶斯的线性模型
x = pd.factorize(SMS_data.factor)[0] # high为0，low为1
# 设置beta1的先验分布的尺度参数，与r=0.707相匹配
beta1_scale = 0.707

# 建立模型
with pm.Model() as linear_regression:
    sigma = pm.HalfCauchy("sigma", beta=2)
    beta0 = pm.Normal("beta0", 0, sigma=5)
    beta1 = pm.Normal("beta1", 0, sigma=beta1_scale)
    # beta1 = pm.StudentT("beta1", nu=1, mu=0, sigma=beta1_scale)  # 使用 StudentT 分布接近 Cauchy 分布
    # beta1 = pm.Cauchy("beta1", alpha=0, beta=beta1_scale)  # 使用 Cauchy 分布，参数 beta 与 r 值对应
    x = pm.Data("x", x)
    mu = pm.Deterministic("μ", beta0 + beta1 * x)
    pm.Normal("y", mu=mu, sigma=sigma, observed=SMS_data.variable)

可以通过pymc自带的可视化工具将模型关系可视化。  

x 为自变量，其中1为低社会关系，0为高社会关系。  

参数 $\beta0$ 是线性模型的截距，而 $\beta1$ 是斜率。  

截距代表了高社会关系地位被试的幸福感；而截距加上斜率表示低社会关系地位被试的幸福感。  

参数$sigma$是残差，因变量$y$即主观幸福感。  

模型图展示了各参数通过怎样的关系影响到因变量。

In [12]:
pm.model_to_graphviz(linear_regression)

Pymc 贝叶斯模型结合 Arviz 计算得到的贝叶斯因子结果(基于 Savage-Dickey 方法)与贝叶斯Ttest结果接近。

In [13]:
# 模型拟合过程 (mcmc采样过程)
with linear_regression:
    # 先验采样
    prior_checks = pm.sample_prior_predictive(samples=5000)
    # 后验采样
    idata = pm.sample(2000, tune=1000, target_accept=0.9, return_inferencedata=True)
    # 将先验采样的结果合并到推断数据中
    idata.extend(pm.sample_prior_predictive(samples=500, model=linear_regression))

Sampling 4 chains, 0 divergences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 / 0:00:09

Sampling 4 chains for 1_000 tune and 2_000 draw iterations (4_000 + 8_000 draws total) took 10 seconds.
Sampling: [beta0, beta1, sigma, y]


In [14]:
fig, axs = plt.subplots(1, 2, figsize=(6, 3))
az.plot_posterior(prior_checks.prior["beta1"], ax=axs[0])
axs[0].set_title("Prior")
az.plot_posterior(idata.posterior["beta1"], ax=axs[1], color="C1")
axs[1].set_title("Posterior")
plt.show()

<Figure size 600x300 with 2 Axes>

In [15]:
print(f"The Bayes T-test is {BF_sms}")
ax = az.plot_bf(idata, var_name="beta1", ref_val=0)
ax[1].set_xlim(-0.5, 0.5)
ax[1].set_xlabel("beta1 \n(represents the difference \nbetween two groups)")
plt.show()

The Bayes T-test is 0.12703440686826656


<Figure size 400x300 with 1 Axes>

### 例2 贝叶斯分层模型（Hierarchical Model） 可以更准确地估计信度  

在前面的内容中，我们通过社会关系地位（高与低）对幸福感的影响进行了可视化，并使用贝叶斯因子和贝叶斯推断对模型进行了进一步的分析。然而，现实世界的数据往往存在分层结构（例如，数据可能来自不同的群体、地区、时间段等），这些层次结构可以对我们感兴趣的效应（如社会关系地位对幸福感的影响）产生影响。  

#### 分层模型（Hierarchical Model）：  
可以自然地扩展我们之前的模型，以捕捉数据中的这种分层结构，从而提供更精细、更具解释力的分析。  

例2的数据来源于Haines (2020)的论文（ https://osf.io/preprints/psyarxiv/xr7y3 ）。  

<p style="font-size:14px">参考文献：</p>  
<p style="font-size:14px">  
Haines, N., Kvam, P. D., Irving, L. H., Smith, C., Beauchaine, T. P., Pitt, M., & Turner, B. (2020, August 24). Theoretically informed generative models can advance the psychological and brain sciences: Lessons from the reliability paradox. PsyArXiv. https://doi.org/10.31234/osf.io/xr7y3  
</p>

例2数据集包括Flanker任务中的反应时间（RT）和正确率数据。Flanker任务要求被试对中央箭头的方向作出反应，同时忽略周围的干扰箭头。  

	•	反应时间 (RT)：以秒为单位，记录每次试验的反应时间。  
	•	正确率 (Correct)：记录被试反应是否正确。  
	•	条件 (Condition)：任务分为一致性和不一致性试验，箭头方向是否一致会影响被试的反应时间。  
	•	试验和区块：任务分为多个区块，每个区块包含一系列试验。  

<div style="text-align: center;">  
    <img src="https://tse1-mm.cn.bing.net/th/id/OIP-C.mWWpZdVKA6TlyCuwQlnQSQHaEs?rs=1&pid=ImgDetMain" alt="Flanker Example" width="640" height="640">  
    <p style="font-size:12px">(Kim et al., 2022)</p>  
</div>

In [16]:
# 导入例2数据
try:
  df_flanker = pd.read_csv('/home/mw/input/bayes3797/flanker_1.csv')
except:
  df_flanker = pd.read_csv('data/flanker_1.csv')

# 分层模型数据预处理
df_model = df_flanker.copy()
df_model = df_model[(df_model.RT > 0) & (df_model.Condition != 1)].reset_index(drop=True)
# 将条件进行重新编码，1为不一致，0为一致
df_model["Condition"] = np.where(df_model["Condition"] == 2, 1,df_model["Condition"])
# 将时间变量进行重新编码
df_model["Time"].replace(1, 0, inplace=True)
df_model["Time"].replace(2, 1, inplace=True)
# 生成所有试次的索引
df_model["trial_idx"] = range(len(df_model))
# 重新编码被试编号
df_model['subj_num'] = df_model['subj_num'] - 1
# 每个被试不同时间rt的最小值
rt_min = df_model.groupby(['subj_num','Time']).RT.min().unstack().to_numpy()

In [17]:
df_model.head()

Block  Trial  Arrow  Condition  Correct       RT  subj_num  Time     Task  \
0      1      1      2          0        1  3.07760         0     0  Flanker   
1      1      2      1          0        1  0.76228         0     0  Flanker   
2      1      3      1          1        1  2.65480         0     0  Flanker   
3      1      4      2          1        1  0.68876         0     0  Flanker   
4      1      5      1          1        1  0.47930         0     0  Flanker   

   trial_idx  
0          0  
1          1  
2          2  
3          3  
4          4

##  <center>Two-stage model <center>

In [18]:
# two-stage model 数据预处理

df_trad = df_flanker.copy()
# 选择反应时大于0且条件不为1
df_trad = df_trad[(df_trad.RT > 0) & (df_trad.Condition != 1)].reset_index(drop=True)
# 将条件进行重新编码，2为不一致，0为一致
df_trad["Condition"] = df_trad["Condition"].replace({2: 'incon', 0: 'con'})
# 将时间变量进行重新编码
df_trad["Time"] = df_trad["Time"].replace({1: 't1', 2: 't2'})
# 根据条件分组
df_trad = df_trad.groupby(['subj_num','Condition','Time'])['RT'].agg(['mean','std']).reset_index()
# 将需要做相关的转成长数据
df_trad= df_trad.pivot(index='subj_num',columns=['Condition','Time'],values=['mean','std']).reset_index()
df_trad.columns = [
    'subj_num' if col1 == 'subj_num' else f'{col1}_{col2}_{col3}' 
    for col1, col2, col3 in df_trad.columns]
# 关注平均值、标准差的差值
df_trad['mean_contrast_t1'] = df_trad['mean_con_t1'] - df_trad['mean_incon_t1']
df_trad['mean_contrast_t2'] = df_trad['mean_con_t2'] - df_trad['mean_incon_t2']
df_trad['sd_contrast_t1'] = df_trad['std_con_t1'] - df_trad['std_incon_t1']
df_trad['sd_contrast_t2'] = df_trad['std_con_t2'] - df_trad['std_incon_t2']
df_trad = df_trad[['subj_num','mean_contrast_t1','mean_contrast_t2','sd_contrast_t1','sd_contrast_t2']]

In [19]:
# 经过数据预处理后，定义函数，计算time 1和time 2数据的皮尔逊相关系数
def pearson_df(method, var, col1, col2): # 定义函数，用于计算皮尔逊相关系数
    contrast_cor = pearsonr(col1, col2) # 计算皮尔逊相关系数
    contrast_coef = contrast_cor[0]
    ci_low, ci_high = contrast_cor.confidence_interval(confidence_level=0.95) # 计算置信区间
    
    # 创建结果数据框
    result_df = pd.DataFrame({
        'mean': [contrast_coef],
        'hdi_2.5%': [ci_low],
        'hdi_97.5%': [ci_high]},
        index = [['Two-Stage'],
                [f'{method}_{var}']])

    return result_df

In [20]:
trad_mu = pearson_df(method = "trad",  
           var = "mean", # 均值
           col1 = df_trad['mean_contrast_t1'], 
           col2 = df_trad['mean_contrast_t2']) # 两个时间点的均值对比

trad_sigma = pearson_df(method = "trad", # 传统方法
           var = "sd", 
           col1 = df_trad['sd_contrast_t1'], 
           col2 = df_trad['sd_contrast_t2'])
# 合并结果
trad_result = pd.concat([trad_mu, trad_sigma]) 
trad_result = trad_result.reset_index() # 重置索引
trad_result["level_1"] = trad_result["level_1"].replace({'trad_mean': 'mu', 'trad_sd': 'sigma'}) # 重命名列名
print(trad_result)

     level_0 level_1      mean  hdi_2.5%  hdi_97.5%
0  Two-Stage      mu  0.317663  0.033557   0.554268
1  Two-Stage   sigma -0.024678 -0.309650   0.264363


### <center>贝叶斯分层模型（Hierarchical Model）<center>  
###  <center>Shifted Lognormal model <center>

In [21]:
def logshift_model(data):
    def shifted_lognormal(mu, sigma, loc, size=None):
        return loc + pm.LogNormal.dist(mu=mu, sigma=sigma, size=size)

    coords = {"cond": data['Condition'].unique(),
            "time": data['Time'].unique(),
            "subj": data['subj_num'].unique(),
            "trial": data['trial_idx']}


    with pm.Model(coords=coords) as model:

        rho_R_mu = pm.LKJCorr('rho_R_mu', n=2, eta=1)
        rho_R_sigma = pm.LKJCorr('rho_R_sigma', n=2, eta=1)

        ######
        triu_idx = np.triu_indices(2, k=1)
        mu_corr_upper = pt.set_subtensor(pt.zeros((2, 2))[triu_idx], rho_R_mu)
        sigma_corr_upper = pt.set_subtensor(pt.zeros((2, 2))[triu_idx], rho_R_sigma)
        R_mu = pt.eye(2) + mu_corr_upper + mu_corr_upper.T
        R_sigma = pt.eye(2) + sigma_corr_upper + sigma_corr_upper.T
        L_R_mu = pt.linalg.cholesky(R_mu)
        L_R_sigma = pt.linalg.cholesky(R_sigma)

        #####

        # Group-level parameter means
        mu_mean_base = pm.Normal('mu_mean_base',mu=0, sigma=1, dims="time")
        mu_sd_base = pm.HalfNormal('mu_sd_base', sigma=1, dims="time")

        
        sigma_mean_base = pm.Normal('sigma_mean_base', mu=0, sigma=1, dims="time")
        sigma_sd_base = pm.HalfNormal('sigma_sd_base', sigma=1, dims="time")

        mu_mean_delta = pm.Normal('mu_mean_delta', mu=0, sigma=1, dims="time")
        mu_sd_delta = pm.HalfNormal('mu_sd_delta', sigma=1, dims="time") #/*

        sigma_mean_delta = pm.Normal('sigma_mean_delta', mu=0, sigma=1, dims="time")
        sigma_sd_delta = pm.HalfNormal('sigma_sd_delta', sigma=1, dims="time") #/*

        ndt_mean = pm.Normal('ndt_mean', mu=0, sigma=1, dims="time")
        ndt_sigma = pm.Normal('ndt_sigma', mu=0, sigma=1, dims="time")

        # to make diagnol matrix/ can multiply with L_R_mu/L_R_sigma
        mu_sd_delta_re = pt.diag(mu_sd_delta)
        sigma_sd_delta_re = pt.diag(sigma_sd_delta)
        L_S_mu =  pm.Deterministic("L_S_mu",pt.dot(mu_sd_delta_re,L_R_mu))
        L_S_sigma = pm.Deterministic("L_S_sigma", pt.dot(sigma_sd_delta_re, L_R_sigma))
        
        # individual-level parameters/congruent
        mu_i_base_pr = pm.Normal('mu_i_base_pr', mu=0, sigma=1, dims=["subj","time"])
        sigma_i_base_pr = pm.Normal("sigma_i_base_pr", mu=0, sigma=1, dims=["subj","time"])
        # individual-level parameters/incongruent
        mu_i_delta_pr = pm.Normal("mu_i_delta_pr", mu=0, sigma=1, dims=["time","subj"])
        sigma_i_delta_pr = pm.Normal("sigma_i_delta_pr", mu=0, sigma=1, dims=["time","subj"])

        ndt_i_pr = pm.Normal('ndt_i_pr', mu=0, sigma=1, dims=["subj","time"])
        rv = pm.Normal.dist(0,1)
        phi = pm.logcdf(rv,ndt_mean+ndt_sigma*ndt_i_pr)
        phi = pm.math.exp(phi)
        ndt_i = pm.Deterministic('ndt_i', phi*rt_min, dims=["subj","time"])

    
        # 得到公式2的后半部分
        mu_i_delta_tilde = pm.Deterministic("mu_i_delta_tilde", pt.dot(L_S_mu, mu_i_delta_pr))
        sigma_i_delta_tilde = pm.Deterministic("sigma_i_delta_tilde", pt.dot(L_S_sigma, sigma_i_delta_pr))

        #----------------------------
        
        # mu congruent
        mu_i_base = pm.Deterministic("mu_i_base", mu_mean_base + mu_sd_base*mu_i_base_pr,dims=["subj","time"])
        # sigma congruent
        sigma_i_base = pm.Deterministic("sigma_i_base", sigma_mean_base + sigma_sd_base*sigma_i_base_pr,dims=["subj","time"])
        # congruent effect
        mu_i_effect = pm.Deterministic("mu_i_effect", mu_mean_delta + mu_i_delta_tilde.T,dims=["subj","time"])
        # sigma incongruent
        sigma_i_effect = pm.Deterministic("sigma_i_effect", sigma_mean_delta + sigma_i_delta_tilde.T,dims=["subj","time"])
        # mu incongruent
        mu_i_incon = pm.Deterministic("mu_i_incon", mu_i_base + mu_i_effect,dims=["subj","time"])
        # sigma incongruent
        sigma_i_incon = pm.Deterministic("sigma_i_incon", sigma_i_base + sigma_i_effect,dims=["subj","time"])

        mu = pm.Deterministic("mu",pm.math.stack([mu_i_base,mu_i_incon]), dims=["cond","subj","time"])
        sigma_stack = pm.Deterministic("sigma_stack",pm.math.stack([sigma_i_base,sigma_i_incon]),dims=["cond","subj","time"])
        sigma = pm.Deterministic("sigma", pm.math.exp(sigma_stack), dims=["cond","subj","time"])

        time_idx = pm.MutableData("time_idx",data.Time, dims="trial")
        subj_idx = pm.MutableData("subj_idx",data.subj_num, dims="trial")
        cond_idx = pm.MutableData("cond_idx",data.Condition, dims="trial")

        likelihood = pm.CustomDist("likelihood",
                    mu[cond_idx, subj_idx, time_idx], 
                    sigma[cond_idx, subj_idx, time_idx],
                    ndt_i[subj_idx, time_idx],
                    dist = shifted_lognormal,
                    observed=data.RT,
                    dims="trial")
                                    
        # pred = pm.Deterministic("pred", likelihood + ndt_i[subj_idx, time_idx], dims="trial")

    return model

In [22]:
# 根据不同的方法选择不同的模型
def process_model(data, distribution):
    if distribution == 'log_shift':
        return logshift_model(data)

In [23]:
# 进行先验预测检验并输出
def prior_predictive_check(model, samples=50, random_seed=43201):
    model_prior_check = pm.sample_prior_predictive(
        samples=samples,
        model=model,
        var_names=["mu","sigma"],
        random_seed=random_seed
    )

    fig, ax = plt.subplots(1, 2, figsize = (15,5))
    fig.suptitle("Prior predictive check")
    ax[0].set_title("mu")
    az.plot_dist(model_prior_check.prior["mu"], ax=ax[0])

    ax[1].set_title("sigma")
    az.plot_dist(model_prior_check.prior["sigma"], ax=ax[1])

    return fig

In [24]:
# 进行采样
def model_sampling(model):
    trace = pm.sample(draws=2000,            # 使用mcmc方法进行采样，draws为采样次数
                    tune=1000,                    # tune为调整采样策略的次数，可以决定这些结果是否要被保留
                    chains=4,                     # 链数
                    cores=4,
                    discard_tuned_samples= True,  # tune的结果将在采样结束后被丢弃
                    random_seed=43201,
                    target_accept=0.8,
                    model=model)
    return trace

In [25]:
# 对重测信度参数进行总结
def diagnostic_summary(trace, model_name):
    diagnostic = az.plot_trace(trace,
                               var_names = ["rho_R_mu", "rho_R_sigma"],
                               figsize=(15,10))
    hdi_sum = az.summary(trace, 
                        var_names =["rho_R_mu", "rho_R_sigma"],
                        hdi_prob = 0.95)
    hdi_sum = hdi_sum[["mean","hdi_2.5%","hdi_97.5%"]]

    hdi_sum.index = pd.MultiIndex.from_arrays([[f'{model_name}'] * len(hdi_sum.index), hdi_sum.index])
    return diagnostic, hdi_sum

In [26]:
# 集合需要的结果并输出
import gc

def result_output(data, distribution, model_name):
    model = process_model(data=data, distribution=distribution)
    prior_plot = prior_predictive_check(model, samples=50, random_seed=43201)
    trace = model_sampling(model)
    diag, hdi = diagnostic_summary(trace, model_name)
    del trace
    gc.collect()
    return diag, hdi, prior_plot

In [27]:
# 使用4核16G的计算资源，运行以下代码，需要约2小时
diag_logshift, hdi_logshift, priorplot_logshift = result_output(df_model, "log_shift","Shifted Lognormal")

Sampling 4 chains, 0 divergences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 / 3:06:42

Sampling 4 chains for 1_000 tune and 2_000 draw iterations (4_000 + 8_000 draws total) took 11203 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details


<Figure size 1500x500 with 2 Axes>

<Figure size 1500x1000 with 4 Axes>

### Two-Stage model vs. Shifted Lognormal model  

与Two-Stage model相比，Shifted Lognormal model具有更高的信度。

In [28]:
# 对变量名进行修改
def rename_hdi(data):
    data = data.reset_index()
    data["level_1"] = data["level_1"].replace({'rho_R_mu[0]': 'mu', 'rho_R_sigma[0]': 'sigma'})
    return data
hdi_logshift = rename_hdi(hdi_logshift)

In [29]:
#选择mu的重测信度参数进行绘制
plot = pd.concat([trad_result, hdi_logshift])
plot = plot[plot['level_1'] == 'mu']
plot

level_0 level_1      mean  hdi_2.5%  hdi_97.5%
0          Two-Stage      mu  0.317663  0.033557   0.554268
0  Shifted Lognormal      mu  0.711000  0.442000   0.952000

In [30]:
plt.figure(figsize=(6,3), dpi=150)

# 计算可信区间
ci = [plot['mean'] - plot['hdi_2.5%'], plot['hdi_97.5%'] - plot['mean']]

# 通过error bar的形式展现区间和均值
plt.errorbar(x=plot['mean'], y=plot['level_0'], xerr=ci,
                color='black', capsize=3, linestyle='None', linewidth=1,
                marker="o", markersize=5, mfc="black", mec="black")

# 调整y轴的刻度使其更集中
plt.gca().set_ylim(-0.5, len(plot['level_0']) - 0.5)  # Adjust based on your data

# 添加标题
plt.xlabel('Test-Retest Correlation', fontsize=8)

sns.despine()

<Figure size 900x450 with 1 Axes>

## 例3: 贝叶斯推断在认知模型中的应用  

在实验数据的收集时，研究者往往会采集个体反应的正确率与反应时。  

而传统分析方法并不能同时对两种数据进行分析，从而推断潜在的认知机制。比如，个体是否愿意牺牲更多的反应时间去获得一个更准确的判断。  


**认知模型能有效的弥补这一问题，比如 drift-diffusion model, DDM。**  


![Image Name](https://cdn.kesci.com/upload/sj8vp25j27.jpeg?imageView2/0/w/960/h/960)  



<center><span style="font-size: 68px;">Part 3: About this course  </span></center>

### 2. 课程内容  
#### 2.0 课时  
9.06 ～ 1.03，18周，十一假期和秋季运动会可能会被放假一次  

#### 2.1 教学目标：  

（1）使用Python进行基本的数据处理  

（2）理解贝叶斯推断的基本原理；  

（2）了解PyMC3的语法和结构，并能应用于相对简单的情境  


### 2.2 考核方式：  

### 考勤(10%) + 3次小作业 (45%) + 大作业 (45%)   

#### 大作业  

大作业要求  

* 真实的数据分析，数据来源：  
  * Science: https://www.science.org/  
  * Science advances: https://www.science.org/journal/sciadv  
  * Nature: https://www.nature.com/  
  * Nature human behaviour：https://www.nature.com/nathumbehav/  
  * Nature communications: https://www.nature.com/ncomms/  
  * Psychological Science: https://journals.sagepub.com/home/pss  
  * Cognition: https://www.sciencedirect.com/journal/cognition  
  * Collabra: Psychology: https://online.ucpress.edu/collabra  
* 合作完成 (每个小组3-5人, 可自行组队或随机分配)  

大作业提交内容  

* 代码 (对公开数据进行完整分析的代码，使用贝叶斯统计方法，具有可重复性，有注释和文档说明)  
	* 如果原文章使用非贝叶斯方法，可以考虑先重复原文档的分析，再进行贝叶斯分析。  
* 与文字报告 (包括标题，摘要，研究背景或研究问题，研究方法，分析结果，结论六个主要部分)  
* 进行汇报 (使用 PPT 汇报对研究问题和代码的分析过程进行汇报)  

大作业评分标准  
* 分工合理、数据分析流程完整、汇报展示清晰美观


### 2.3 课程风格：  
	（1）内容有挑战、考核不复杂  
	（2）1/3一节课展示或互动抄代码  
	（3）专门设有答疑时间，助教给大家答疑解惑  





### 2.4 课程大纲  
### Intro  
1: 课程介绍（为什么要用贝叶斯/PyMC3，展示一个贝叶斯因子和贝叶斯层级模型例子，课程安排）  

### I Basics：  
2: 贝叶斯公式  
* 单一事件的贝叶斯模型(先验、似然、分母和后验)  
* 随机变量的贝叶斯模型(先验、似然、分母和后验)  

3: 建立一个完成的贝叶斯模型：Beta-Binomial model  
* Beta先验  
* Binomial与似然  
* Beta-Binomial model  

4: 贝叶斯模型的特点：数据-先验与动态更新  
* 先验与数据对后验的影响  
* 数据顺序不影响后验  

5: 经典的贝叶斯模型：Conjugate families  
* Gamma-Poisson  
* Normal-Normal  

### II 近似后验估计  
6: 近似后验的方法  
* 网格法  
* MCMC  

7: 深入一种MCMC算法  
* M-H 算法  

8: 基于后验的推断  
* 后验估计  
* 假设检验  
* 后验预没  

### III Bayesian回归模型  
9: 简单的Bayesian线性回归模型  
* 建立模型  
* 调整先验  
* 近似后验  
* 基于近似后验的推断  
* 序列分析  

10: 对回归模型的评估  
* 评估模型的标准  
* 对简单线性模型的估计  
  
11: 扩展的线性模型  
* 多自变量的线性回归  

12: GLM: Bayes factors  

13: GLM: Logistic Regression  

14: Hierachical Bayesian Model 1  

15: Hierachical Bayesian Model 2

### 2.5 参考书  

#### 1. 英文教材：  

![Image Name](https://cdn.kesci.com/upload/sj8izbv81l.png?imageView2/0/w/960/h/960)  


#### 2. 中文教材：  

![Image Name](https://cdn.kesci.com/upload/sj8iytode2.png?imageView2/0/w/320/h/320)  


### 2.6 课程好帮手（AI工具）  

- 如果你有问题，除了可以向老师和助教寻求帮助，还可以使用AI工具。  
- 这款工具是为贝叶斯课程特别定制的，免费且无需翻墙。  

<p align="center">  
  <img src="https://cdn.kesci.com/upload/sj6zpgijv9.jpeg?imageView2/0/w/960/h/960" alt="Image Name" />  
</p>

<center><span style="font-size: 68px;">THANK YOU  </span></center>